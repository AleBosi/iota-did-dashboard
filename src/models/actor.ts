// /models/actor.ts

export interface Actor {
  id: string; // DID unico (es: "did:iota:evm:0x...")
  name: string;
  role: "admin" | "azienda" | "creator" | "operatore" | "macchinario";
  aziendaId?: string;      // DID dell’azienda di appartenenza
  seed?: string;           // non esportare mai in VC!
  publicKey?: string;      // per interoperabilità/validazione
  createdAt?: string;      // opzionale, ISO date
  updatedAt?: string;      // opzionale

  // AGGIUNGI QUESTO CAMPO:
  vcIds?: string[];        // Lista degli ID delle VC associate a questo attore
}
