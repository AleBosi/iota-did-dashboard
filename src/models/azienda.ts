// /models/azienda.ts
export interface Azienda {
  id: string;         // DID (es: did:iota:evm:...)
  name: string;
  seed: string;       // per login e firma
  legalInfo?: {
    vat?: string;         // partita IVA o VAT
    lei?: string;         // Legal Entity Identifier (EBSI ready)
    address?: string;
    email?: string;
    country?: string;
    [key: string]: any;   // altri campi liberi
  };
  creators: string[];     // array di DID
  operatori: string[];
  macchinari: string[];
  createdAt?: string;     // ISO
  updatedAt?: string;     // ISO
}
