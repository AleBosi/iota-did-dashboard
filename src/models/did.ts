// /models/did.ts
export interface DidDocument {
  id: string; // es: "did:iota:evm:0x123..."
  controller?: string[];
  verificationMethod?: {
    id: string;
    type: string;
    publicKeyMultibase: string;
  }[];
  authentication?: string[];
  service?: {
    id: string;
    type: string;
    serviceEndpoint: string;
  }[];
  // Altri campi secondo specifica, se serve
}
