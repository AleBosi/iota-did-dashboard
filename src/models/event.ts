// /models/event.ts

export interface Event {
  id: string;
  productId: string;         // ID prodotto collegato (ora obbligatorio)
  operatoreId: string;       // ID operatore coinvolto
  macchinarioId: string;     // ID macchinario coinvolto
  type: string;              // es: "Assemblaggio", "Controllo", "Assegnazione"
  description: string;
  date: string;              // ISO
  creatorId: string;         // DID del creator che ha creato lâ€™evento
  done?: boolean;            // true se eseguito
  bomComponent?: string;     // opzionale, componente specifico
  proofId?: string;          // se firmerai evento come VC
  vcIds?: string[];          // VC associate/notarizzate collegate a questo evento
}
