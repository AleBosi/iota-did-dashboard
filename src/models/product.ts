// /models/product.ts
export interface Product {
  id: string;              // ID univoco locale/app
  did: string;             // DID prodotto (es: did:iota:evm:0x...)
  name: string;            // Nome commerciale o etichetta prodotto
  typeId?: string;         // riferimento a ProductType.id
  serial?: string;         // seriale prodotto
  owner?: string;          // DID del proprietario corrente
  bom?: { id: string; name: string }[]; // lista componenti (Bill of Materials)
  credentials?: string[];  // lista di VC associate (ID delle VC)
  description?: string;    // opzionale
}
