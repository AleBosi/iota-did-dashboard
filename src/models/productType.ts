// /models/productType.ts
export interface ProductType {
  id: string;               // es: GTIN, oppure UUID univoco
  name: string;
  description?: string;
}
