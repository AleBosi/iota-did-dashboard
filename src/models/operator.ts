// /models/operator.ts
import { Actor } from "./actor";

export interface Operator extends Actor {
  // aggiungi campi solo se servono SOLO agli operatori
}
