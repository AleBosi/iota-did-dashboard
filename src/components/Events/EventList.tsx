import { useState } from "react";
import { Event } from "../../models/event";
import { Actor } from "../../models/actor";
import EventDetails from "./EventDetails";

interface Props {
  events: Event[];
  actors: Actor[]; // serve per mostrare nomi/did nella lista e nei dettagli!
  onSelect?: (event: Event) => void;
}

export default function EventList({ events, actors, onSelect }: Props) {
  const [selected, setSelected] = useState<Event | null>(null);

  // Helper per trovare attore per id
  const findActor = (id?: string) => actors.find(a => a.id === id);

  return (
    <div className="flex gap-8">
      <ul className="w-1/2">
        {events.map(ev => (
          <li
            key={ev.id}
            className={`border-b py-2 cursor-pointer hover:bg-blue-50 ${ev.done ? "line-through text-gray-400" : ""}`}
            onClick={() => {
              setSelected(ev);
              onSelect?.(ev);
            }}
          >
            <span className="font-semibold">{ev.type}</span>
            {ev.bomComponent && (
              <span className="ml-2 text-gray-600">[{ev.bomComponent}]</span>
            )}
            <span className="ml-2">{ev.description}</span>
            {/* Operatore */}
            {ev.operatoreId && (
              <span className="ml-2 text-xs text-blue-800">
                üë§ {findActor(ev.operatoreId)?.name ?? ev.operatoreId}
              </span>
            )}
            {/* Macchinario */}
            {ev.macchinarioId && (
              <span className="ml-2 text-xs text-purple-800">
                üè≠ {findActor(ev.macchinarioId)?.name ?? ev.macchinarioId}
              </span>
            )}
            <span className="ml-2 text-xs text-gray-400">
              {new Date(ev.date).toLocaleString()}
            </span>
            {ev.done && (
              <span className="ml-2 text-green-600 font-bold">‚úì</span>
            )}
          </li>
        ))}
      </ul>
      <div className="w-1/2">
        {selected && <EventDetails event={selected} actors={actors} />}
      </div>
    </div>
  );
}
