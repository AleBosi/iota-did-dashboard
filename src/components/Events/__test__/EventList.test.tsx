import { render, screen, fireEvent } from "@testing-library/react";
import EventList from "../EventList";
import { Event } from "../../../models/event";
import { Actor } from "../../../models/actor";

// Mocks
const actors: Actor[] = [
  {
    id: "op-1",
    did: "did:example:op-1",
    name: "Anna Op",
    role: "operatore",
  },
  {
    id: "mac-1",
    did: "did:example:mac-1",
    name: "Mac1",
    role: "macchinario",
  },
  {
    id: "op-2",
    did: "did:example:op-2",
    name: "Bob Op",
    role: "operatore",
  },
  {
    id: "mac-2",
    did: "did:example:mac-2",
    name: "Mac2",
    role: "macchinario",
  },
];

const mockEvents: Event[] = [
  {
    id: "evt-1",
    type: "avvio",
    description: "Avvio macchina A",
    productId: "prod-1",
    operatoreId: "op-1",
    macchinarioId: "mac-1",
    creatorId: "did:admin",
    date: "2025-08-01T10:00:00Z",
    done: false,
  },
  {
    id: "evt-2",
    type: "manutenzione",
    description: "Sostituzione filtro",
    productId: "prod-2",
    operatoreId: "op-2",
    macchinarioId: "mac-2",
    creatorId: "did:admin",
    date: "2025-08-02T15:30:00Z",
    bomComponent: "filtro-aria",
    done: true,
  },
];

describe("EventList component", () => {
  it("renderizza tutti gli eventi come lista", () => {
    render(<EventList events={mockEvents} actors={actors} />);
    expect(screen.getByText("avvio")).toBeInTheDocument();
    expect(screen.getByText("Avvio macchina A")).toBeInTheDocument();
    expect(screen.getByText("manutenzione")).toBeInTheDocument();
    expect(screen.getByText("Sostituzione filtro")).toBeInTheDocument();
    expect(screen.getByText((t) => t.includes("Anna Op"))).toBeInTheDocument();
    expect(screen.getByText((t) => t.includes("Mac1"))).toBeInTheDocument();
    expect(screen.getByText((t) => t.includes("Bob Op"))).toBeInTheDocument();
    expect(screen.getByText((t) => t.includes("Mac2"))).toBeInTheDocument();
  });

  it("visualizza bomComponent se presente", () => {
    render(<EventList events={mockEvents} actors={actors} />);
    expect(screen.getByText("[filtro-aria]")).toBeInTheDocument();
  });

  it("applica stile line-through e ✓ se done è true", () => {
    render(<EventList events={mockEvents} actors={actors} />);
    const manutenzioneLi = screen.getByText("manutenzione").closest("li");
    expect(manutenzioneLi).toHaveClass("line-through");
    expect(screen.getByText("✓")).toBeInTheDocument();
  });

  it("mostra EventDetails al click su un evento", () => {
    render(<EventList events={mockEvents} actors={actors} />);
    fireEvent.click(screen.getByText("avvio"));
    expect(screen.getByText(/ID evento:/i)).toBeInTheDocument();
    expect(screen.getByText("evt-1")).toBeInTheDocument();
    // Usa getAllByText: può essere presente sia nella lista che nei dettagli
    expect(screen.getAllByText((t) => t.includes("Anna Op")).length).toBeGreaterThan(0);
    expect(screen.getAllByText((t) => t.includes("Mac1")).length).toBeGreaterThan(0);
  });

  it("chiama onSelect quando un evento viene selezionato", () => {
    const onSelect = jest.fn();
    render(<EventList events={mockEvents} actors={actors} onSelect={onSelect} />);
    fireEvent.click(screen.getByText("manutenzione"));
    expect(onSelect).toHaveBeenCalledWith(mockEvents[1]);
  });

  it("snapshot", () => {
    const { asFragment } = render(<EventList events={mockEvents} actors={actors} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
