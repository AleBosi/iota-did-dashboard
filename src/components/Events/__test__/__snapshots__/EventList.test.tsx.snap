// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`EventList component snapshot 1`] = `
<DocumentFragment>
  <div
    class="flex gap-8"
  >
    <ul
      class="w-1/2"
    >
      <li
        class="border-b py-2 cursor-pointer hover:bg-blue-50 "
      >
        <span
          class="font-semibold"
        >
          avvio
        </span>
        <span
          class="ml-2"
        >
          Avvio macchina A
        </span>
        <span
          class="ml-2 text-xs text-blue-800"
        >
          👤 Anna Op
        </span>
        <span
          class="ml-2 text-xs text-purple-800"
        >
          🏭 Mac1
        </span>
        <span
          class="ml-2 text-xs text-gray-400"
        >
          8/1/2025, 12:00:00 PM
        </span>
      </li>
      <li
        class="border-b py-2 cursor-pointer hover:bg-blue-50 line-through text-gray-400"
      >
        <span
          class="font-semibold"
        >
          manutenzione
        </span>
        <span
          class="ml-2 text-gray-600"
        >
          [filtro-aria]
        </span>
        <span
          class="ml-2"
        >
          Sostituzione filtro
        </span>
        <span
          class="ml-2 text-xs text-blue-800"
        >
          👤 Bob Op
        </span>
        <span
          class="ml-2 text-xs text-purple-800"
        >
          🏭 Mac2
        </span>
        <span
          class="ml-2 text-xs text-gray-400"
        >
          8/2/2025, 5:30:00 PM
        </span>
        <span
          class="ml-2 text-green-600 font-bold"
        >
          ✓
        </span>
      </li>
    </ul>
    <div
      class="w-1/2"
    />
  </div>
</DocumentFragment>
`;
